libzmix/src/signatures/delg_cred_cdd/issuer.rs

```diff
@@ -0,0 +1,1066 @@
+use super::errors::{DelgError, DelgResult};
+use super::groth_sig::{
+    Groth1SetupParams, Groth1Sig, Groth1Verkey, Groth2SetupParams, Groth2Sig, Groth2Verkey,
+    GrothS1, GrothS2, GrothSigkey,
+};
+use amcl_wrapper::group_elem::{GroupElement, GroupElementVector};
+use amcl_wrapper::group_elem_g1::{G1LookupTable, G1Vector, G1};
+use amcl_wrapper::group_elem_g2::{G2Vector, G2};
+
+pub type Sigkey = GrothSigkey;
+pub type EvenLevelVerkey = Groth1Verkey;
+pub type OddLevelVerkey = Groth2Verkey;
+
+// (attributes, signature). The signature is over the attributes and the public key combined by appending public key to the attribute vector.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CredLinkOdd {
+    pub level: usize,
+    pub attributes: G1Vector,
+    pub signature: Groth1Sig,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CredLinkEven {
+    pub level: usize,
+    pub attributes: G2Vector,
+    pub signature: Groth2Sig,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CredChain {
+    pub odd_links: Vec<CredLinkOdd>,
+    pub even_links: Vec<CredLinkEven>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct EvenLevelIssuer {
+    pub level: usize,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct OddLevelIssuer {
+    pub level: usize,
+}
+
+pub struct RootIssuer {}
+
+pub type RootIssuerVerkey = EvenLevelVerkey;
+
+impl CredLinkOdd {
+    pub fn attribute_count(&self) -> usize {
+        self.attributes.len()
+    }
+
+    pub fn has_verkey(&self, vk: &OddLevelVerkey) -> bool {
+        self.attributes[self.attributes.len() - 1] == vk.0
+    }
+
+    pub fn verify(
+        &self,
+        delegatee_vk: &OddLevelVerkey,
+        delegator_vk: &EvenLevelVerkey,
+        setup_params: &Groth1SetupParams,
+    ) -> DelgResult<bool> {
+        if self.attributes.len() > setup_params.y.len() {
+            return Err(DelgError::MoreAttributesThanExpected {
+                expected: setup_params.y.len(),
+                given: self.attributes.len(),
+            });
+        }
+        if !self.has_verkey(delegatee_vk) {
+            return Err(DelgError::VerkeyNotFoundInDelegationLink {});
+        }
+        /*link.signature
+        .verify(link.messages.as_slice(), delegator_vk, setup_params)*/
+        self.signature
+            .verify_fast(self.attributes.as_slice(), delegator_vk, setup_params)
+    }
+}
+
+impl CredLinkEven {
```

Comment:
I don't see a difference between `CredLinkOdd` and `CredLinkEven`. Is the purpose strictly for typing?

Reply:
The members have different types